### ISSUE TITLE
pkg: add key-sorted versions of {json,yaml}.Marshal{,Stream}

### ISSUE THREAD

**Issue by kshpytsya** [cue-lang/cue#3595]
**Is your feature request related to a problem? Please describe.**
My use of cuelang is in context of a complex (currently 2.5Kloc and growing) multi-environment configuration which undergoes frequent refactorings as new features are added. The major artifact of configuration build pipeline is a set of flux yaml files which are pushed to dedicated branches. Changes to protected (think production) environments cause PRs to be raised for mandatory review. And here lies the problem: quite often, refactoring of cue, which is meant to keep artifacts unchanged, causes the order of keys in maps to change which causes garbage PRs to be raised causing wasted time of reviewers. And even without PRs, it causes "garbage" changes in the git history of generated flux.

**Describe the solution you'd like**
I would like to have versions of `{json,yaml}.Marshal{,Stream}` functions that would emit objects with keys in lexicographical order.

**Describe alternatives you've considered**
If this feature doesn't land cuelang library soon I will be forced to introduce some kind of post-processing of yaml files that would order the keys. The problem with this approach aside of introduction of yet another tool into the pipeline (and local development environment) is that sometimes there are nested structures, like json in yaml ConfigMap or even deeper, and it would be intractable to implement proper key-sorting as post-processing.

**Comment by cuematthew** [comment]
Hi @kshpytsya. Thanks for your report.

I think we have the best news: we have already implemented what you're after.

If you're using 0.11, there is an experiment called toposort. There are some words about what it's trying to do here: https://github.com/cue-lang/cue/issues/3558 With this experiment enabled (via the CUE_EXPERIMENT env var - there's no Go API to enable it I'm afraid) you should find it applies to not only `cue eval` and `cue export`, and not only the Go API around [Value.Fields](https://pkg.go.dev/cuelang.org/go/cue#Value.Fields), but also the CUE stdlib `{json,yaml}.Marshal{,Stream}` functions.

There is however a good chance that toposort won't be sufficient for you: there are a number of circumstances where source-code-position does end up influencing the ordering, but I do suggest you try it first.

If toposort isn't good enough and you want a full lexicographical sort, then you need to be running the tip of master: we added a lexicographical sort, and got it working ... yesterday. To turn this on, I'm afraid you need a different env var: `CUE_DEBUG=sortfields`. This takes precedence over `CUE_EXPERIMENT=toposort`. But just like with toposort, it should work for `cue eval`, `cue export`, the Go API, and the CUE stdlib.

Given this is somewhat bleeding edge, there's a chance we've overlooked something (there are known differences for toposort between the different versions of the evaluator; and there are rumours of a bug with the lexico sort too), so please do let us know how you get on.

**Comment by kshpytsya** [comment]
Thank you for your swift response. Toposort is indeed not good enough for my needs as refactorings may indeed drastically change evaluation order while not changing the results (as evaluated using key-order insensitive comparison of maps). Moreover what I am looking for is not "good enough" but "correct". And in this case, correct is lexicographical order.

We use https://github.com/cue-lang/setup-cue in our pipeline. Is there any specific version I can specify that would include the  `CUE_DEBUG=sortfields` feature?

**Comment by cuematthew** [comment]
Having had a quick look through the code for setup-cue, it looks to me like this won't work - the "latest" there is the latest released version, which is currently 0.11, and not the tip of master. Not sure what to suggest here - I'll ask around.

**Comment by cuematthew** [comment]
Ok, I'm afraid we don't currently have any work scheduled to extend setup-cue. So unfortunately, the lexical sort won't available via setup-cue until we release a 0.12-alpha, which might be a few weeks away, sorry.

**Comment by kshpytsya** [comment]
A few weeks is ok with me. Will this feature be behind the `CUE_DEBUG` flag? It sounds a bit off-putting, as if it is not really supposed to be normally used this way. I am somewhat surprised why this has to be a global feature-flag rather than just  a new separate function. Or, maybe even something like a separate function that would recursively sort keys so that one would write `yaml.Marshal(whatever.RecursiveSortKeys(object))`. But hey, not my realm, so I will use any solution that I can put to practical use.
P.S. I can envision that a single project may want both order-preserving (i.e. as specified in source text) and sorted outputs for different use-cases.

**Comment by cuematthew** [comment]
Initially yes, it will be behind `CUE_DEBUG`, and yes I agree it's not a great place for it. Personally, I'm never a fan of configuring any program via env vars, no matter what they're called. I think there is widespread agreement that it shouldn't remain there given it is so broadly useful. There is some thinking that it should be exposed as an encoder option (e.g. `cue export json+sortfields: foo.cue`). All of this work is fairly new and will take some time to bed down.

The other way of looking at this is that JSON at least (I'm not 100% sure about YAML) defines an object as "a collection of name/value pairs". I.e. unordered. Consequently, you could make a pretty sane argument that any time you're comparing two json objects, you should be ignoring the order of the fields in those objects (i.e. imposing some other order, perhaps lexicographical, as part of the diff-calculation/viewing). Unfortunately, github for one, does not seem to support calculating diffs of files of JSON with this in mind. There is tooling out there that does ignore field order of the inputs, but as you say in your first post, that's yet another tool to integrate, manage, maintain etc.

**Comment by mvdan** [comment]
Just to add my 2c: you don't need to set an env var globally, you can use the context option at https://pkg.go.dev/cuelang.org/go/cue/cuecontext#CUE_DEBUG.

**Comment by kshpytsya** [comment]
> Just to add my 2c: you don't need to set an env var globally, you can use the context option at https://pkg.go.dev/cuelang.org/go/cue/cuecontext#CUE_DEBUG.

Irrelevant, as I am using cue as a cli tool from bash wrapper scripts.

**Comment by mvdan** [comment]
Fair enough, I'm just giving more options, as you hadn't specified whether you were using the CLI or the Go API.

We will explore ways to turn on this option in a more fine-grained way in the future, but as Matthew mentioned, for now we're trying to roll out a global option and flesh out its bugs.

**Comment by mvdan** [comment]
Re-reading this thread now - @kshpytsya I assume you meant this feature request for the APIs which can be imported and used from CUE files, e.g. https://pkg.go.dev/cuelang.org/go/pkg/encoding/json, and not the Go APIs which can be used from Go, like https://pkg.go.dev/cuelang.org/go/encoding/json.
